### PR TITLE
DEV-2178: Refactor Org Plans

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### What's this PR do?

This PR refactors our approach to organization plans, retiring the `Feature` and `Plan` models, adding a `.plan_name` field to `Organization` that points to one two pre-defined `Plan` dataclasses: `free_plan`, and `plus_plan`. These plans correspond to the Free and Plus plans as described [here](https://fundjournalism.org/news-revenue-engine/). Note that in short term, we only need to support those two options even though the marketing materials refer to another plan called "Core".

As specified [in the ticket](https://news-revenue-hub.atlassian.net/jira/software/c/projects/DEV/boards/3?modal=detail&selectedIssue=DEV-2178), this ticket is meant to only encompass the data model changes, and plan-based restrictions are to be handled in follow up tickets. 

One exception to this is that I did go ahead and reimplement plan-based page limits. As you'll see, this involved moving this validation _off_ of the model layer and _onto_ the serializer layer, which is where it should be enforced. I created a [follow up ticket](https://news-revenue-hub.atlassian.net/browse/DEV-2434) to add a confirmation dialog to the Django admin when page creation would lead to an Org having more pages than allowed by its plan. For now, this won't be enforced though.  In contrast, the SPA will be prevented from triggering new page creation if an org is at its limit with the current changes.

Specifics:

- Removes all existing tests and references to `Plan` and `Feature` models
- Adds a single migration (apps/organizations/migrations/0008_DEV-2178_refactor_plan_modeling_and_migrate_legacy_clients.py) that adds `Organization.page_name` (defaulting to the free plan's name), migrates legacy clients to PLUS plan, and deletes the `Plan` and `Feature` models and their references.
- Adds a `Plan` dataclass that is for modeling org plans, but without the database layer being involved. There are two instances of `Plan`: `free_plan` and `plus_plan`.

 
#### Why are we doing this? How does it help us?

This is a requirement for self-onboarding.

#### How should this be manually tested? Please include detailed step-by-step instructions.

Will be easier to manually test in subsequent tickets that  restrict access to SPA-side functionality based on org plan. 

In short term, you can verify that the page limit functionality works by logging in to the dashboard and trying to add a page for an org that already has one page. For example, try to add a new page for Billypenn [here](https://dev-2178.revengine-review.org/pages). 

When you do, you should see an error message stating the org has no pages left.

Screen Shot 2022-09-22 at 9 58 06 AM

Next, go to NRH admin [here](https://dev-2178.revengine-review.org/nrhadmin/organizations/organization/2/change/) and change Billypenn's plan to "PLUS". Now go back to SPA and try again to add a new page. This time it should be successful.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No

#### Have automated unit tests been added? If not, why?

Yes

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No

#### Has this been documented? If so, where?

N/A

#### What are the relevant tickets? Add a link to any relevant ones.

[DEV-2178](https://news-revenue-hub.atlassian.net/jira/software/c/projects/DEV/boards/3?modal=detail&selectedIssue=DEV-2178)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

In addition to other tickets already created for the parent epic, I created [DEV-2434](https://news-revenue-hub.atlassian.net/browse/DEV-2434) for next steps around adding confirmation view to Django admin when creating plan-limited entities.

### CODE REVIEW

**Review by njharman — CHANGES_REQUESTED** [review]
Good test coverage.

File: apps/organizations/migrations/0008_DEV-2178_refactor_plan_modeling_and_migrate_legacy_clients.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+]
+
+
+def migrate_legacy_clients_to_plus_plan(apps, schema_editor):
+    """This migrates existing legacy client orgs to the PLUS plan, ahead of remove feature and plan
+    models.
+    """
+    # RevenueProgram = apps.get_model("organizations", "RevenueProgram")
+    Organization = apps.get_model("organizations", "Organization")
+    # orgs = RevenueProgram.objects.filter(slug__in=rp_slugs_for_legacy_clients).values_list("organization").distinct()
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/organizations/migrations/0008_DEV-2178_refactor_plan_modeling_and_migrate_legacy_clients.py]: Missed that. Yes, stale.

3d51e6451

File: apps/organizations/models.py [hunk 2]
```diff
@@ -15,66 +17,66 @@
 from apps.users.models import RoleAssignmentResourceModelMixin, UnexpectedRoleType
 
 
+FAKE_INFINITY = 200
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/organizations/models.py]: a38b51646e

File: apps/organizations/models.py [hunk 3]
```diff
@@ -15,66 +17,66 @@
...
 RP_SLUG_MAX_LENGTH = 63
 
 
-class Feature(IndexedTimeStampedModel):
-    VALID_BOOLEAN_INPUTS = ["t", "f", "0", "1"]
-
-    class FeatureType(models.TextChoices):
-        PAGE_LIMIT = "PL", ("Page Limit")
-        BOOLEAN = "BL", ("Boolean")
+CURRENCY_CHOICES = [(k, k) for k, _ in settings.CURRENCIES.items()]
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/organizations/models.py]: 06b2b8a1

File: apps/organizations/models.py [hunk 4]
```diff
@@ -15,66 +17,66 @@
...
-        if role_assignment.role_type == Roles.HUB_ADMIN:
-            return queryset.all()
-        elif role_assignment.role_type in (Roles.ORG_ADMIN, Roles.RP_ADMIN):
-            return queryset.filter(organization=role_assignment.organization)
-        else:
-            raise UnexpectedRoleType(f"{role_assignment.role_type} is not a valid value")
+class Plans(models.TextChoices):
 
+    FREE = free_plan.name, _(free_plan.label)
+    PLUS = plus_plan.name, _(plus_plan.label)
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/organizations/models.py]: f67d147a8

File: apps/organizations/models.py [hunk 5]
```diff
@@ -15,66 +17,66 @@
...
 
+free_plan = Plan(
+    name="FREE",
+    label="Free",
+)
 
-class Plan(IndexedTimeStampedModel, RoleAssignmentResourceModelMixin):
-    name = models.CharField(max_length=255)
-    features = models.ManyToManyField("organizations.Feature", related_name="plans", blank=True)
+plus_plan = Plan(
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/organizations/models.py]: f67d147a8

File: apps/organizations/models.py [hunk 6]
```diff
@@ -15,66 +17,66 @@
 from apps.users.models import RoleAssignmentResourceModelMixin, UnexpectedRoleType
 
 
+FAKE_INFINITY = 200
+
 logger = logging.getLogger(f"{settings.DEFAULT_LOGGER}.{__name__}")
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/organizations/models.py:19]: 04a6a1106

File: apps/organizations/migrations/0008_DEV-2178_refactor_plan_modeling_and_migrate_legacy_clients.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
+# Generated by Django 3.2.15 on 2022-09-20 14:24
+from django.db import migrations, models
+
+# from https://news-revenue-hub.atlassian.net/wiki/spaces/TECH/pages/2214625287/Existing+Client+Donation+Page+URLs
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/organizations/migrations/0008_DEV-2178_refactor_plan_modeling_and_migrate_legacy_clients.py:4]: f893007a8

File: apps/organizations/models.py [hunk 8]
```diff
@@ -15,66 +17,66 @@
...
-            raise UnexpectedRoleType(f"{role_assignment.role_type} is not a valid value")
+class Plans(models.TextChoices):
 
+    FREE = free_plan.name, _(free_plan.label)
+    PLUS = plus_plan.name, _(plus_plan.label)
 
-CURRENCY_CHOICES = [(k, k) for k, _ in settings.CURRENCIES.items()]
+    @classmethod
+    def get_plan(cls, name):
+        return ({Plans.FREE.value: free_plan, Plans.PLUS.value: plus_plan}).get(name, None)
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/organizations/models.py]: f70fa261c2af

[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/organizations/models.py]: Resolved in one of several cleanup commits. Didn't catch the has and don't want to dig. As you can see, it's outdated.

[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/organizations/models.py]: f70fa261c2af

**Review by njharman — APPROVED** [review]

File: apps/organizations/migrations/0008_DEV-2178_refactor_plan_modeling_and_migrate_legacy_clients.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+# as of 9/19/22
+rp_slugs_for_legacy_clients = [
```
Comment (x110dc) [inline apps/organizations/migrations/0008_DEV-2178_refactor_plan_modeling_and_migrate_legacy_clients.py:6]: I don't like hardcoding this in the migration file. We can handle this as a one-off ad-hoc script.

File: apps/organizations/models.py [hunk 10]
```diff
@@ -17,64 +18,67 @@
...
-        blank=False,
-        help_text="Limit feature types must be a positive integer. Valid Boolean Type values are ('t', 'f', '1', '0')",
-    )
-    description = models.TextField(blank=True)
+    name: str
+    label: str
+    analytics_enabled: bool = False
+    custom_fonts_enabled: bool = False
+    custom_reports_enabled: bool = False
+    custom_thank_you_pages_enabled: bool = False
```
Comment (x110dc) [inline apps/organizations/models.py:44]: Not sure of the pros/cons here, but is there value in only adding attributes here after we have a reason to implement them? E.g. should we wait to add `custom_fonts_enabled` here when we implement that limit?

**Review by x110dc — CHANGES_REQUESTED** [review]
Requested a couple changes. Ping me and we can discuss the ad-hoc migration script.

File: apps/organizations/models.py (hunk 10 above)
Comment (benjaminEwhite) [reply apps/organizations/models.py:44]: Yeah, I agree, and was on fence. First pass at this was based on looking at the pricing page. I only subsequently looked at tickets for dealing with individual portions. 

I think I'll pare down to just what's essential for this PR.

File: apps/organizations/migrations/0008_DEV-2178_refactor_plan_modeling_and_migrate_legacy_clients.py (hunk 9 above)
Comment (benjaminEwhite) [reply apps/organizations/migrations/0008_DEV-2178_refactor_plan_modeling_and_migrate_legacy_clients.py:6]: Okay. I'm breaking this current single PR into two separate ones, with the intention that we'll run the ad hoc script after merging first and before merging second. Shared a link to a Gist that has script in JIRA ticket.

Comment (benjaminEwhite) [reply apps/organizations/migrations/0008_DEV-2178_refactor_plan_modeling_and_migrate_legacy_clients.py]: In  ab198d2cef,  removed hard coded data migration and saved as a Gist, which I linked to [in the JIRA issue](https://news-revenue-hub.atlassian.net/browse/DEV-2178).

File: apps/organizations/models.py (hunk 10 above)
Comment (benjaminEwhite) [reply apps/organizations/models.py]: 908bf98e

**Review by x110dc — APPROVED** [review]

**Pull request merged**